src/test/java/com/squareup/javapoet/TypesEclipseTest.java

```diff
@@ -0,0 +1,158 @@
+public final class TypesEclipseTest extends AbstractTypesTest {
+  /**
+   * A {@link JUnit4} {@link Rule} that executes tests such that a instances of {@link Elements} and
+   * {@link Types} are available during execution.
+   *
+   * <p>To use this rule in a test, just add the following field: <pre>   {@code
+   *   @Rule public CompilationRule compilationRule = new CompilationRule();}
+   *
+   * @author Gregory Kick
+   */
+  public static final class CompilationRule implements TestRule {
+    private Elements elements;
+    private Types types;
+
+    @Override
+    public Statement apply(final Statement base, Description description) {
+      return new Statement() {
+        @Override public void evaluate() throws Throwable {
+          final AtomicReference<Throwable> thrown = new AtomicReference<Throwable>();
+          boolean successful = compile(ImmutableList.of(new AbstractProcessor() {
+            @Override
+            public SourceVersion getSupportedSourceVersion() {
+              return SourceVersion.latest();
+            }
+
+            @Override
+            public Set<String> getSupportedAnnotationTypes() {
+              return ImmutableSet.of("*");
+            }
+
+            @Override
+            public synchronized void init(ProcessingEnvironment processingEnv) {
+              super.init(processingEnv);
+              elements = processingEnv.getElementUtils();
+              types = processingEnv.getTypeUtils();
+            }
+
+            @Override
+            public boolean process(Set<? extends TypeElement> annotations,
+                RoundEnvironment roundEnv) {
+              // just run the test on the last round after compilation is over
+              if (roundEnv.processingOver()) {
+                try {
+                  base.evaluate();
+                } catch (Throwable e) {
+                  thrown.set(e);
+                }
+              }
+              return false;
+            }
+          }));
+          checkState(successful);
+          Throwable t = thrown.get();
+          if (t != null) {
+            throw t;
+          }
+        }
+      };
+    }
+
+    /**
+     * Returns the {@link Elements} instance associated with the current execution of the rule.
+     *
+     * @throws IllegalStateException if this method is invoked outside the execution of the rule.
+     */
+    public Elements getElements() {
+      checkState(elements != null, "Not running within the rule");
+      return elements;
+    }
+
+    /**
+     * Returns the {@link Types} instance associated with the current execution of the rule.
+     *
+     * @throws IllegalStateException if this method is invoked outside the execution of the rule.
+     */
+    public Types getTypes() {
+      checkState(elements != null, "Not running within the rule");
+      return types;
+    }
+
+    static private boolean compile(Iterable<? extends Processor> processors) {
+      JavaCompiler compiler = new EclipseCompiler();
```

Comment:
Occasionally I try to take a stab at making compile testing work with `ecj`, but end up running into walls where it makes assumptions about the implementation of the filer and whatnot and ends up crashing.  I haven't tried in particular with `CompilationRule` though.  It might work, but I wouldn't be remotely surprised if it didn't either.